specs/PermissionManagement.md

```diff
@@ -0,0 +1,561 @@
+    auto extendedArgs = args.try_query<ICoreWebView2PermissionRequestedEventArgs3>();
+    if (extendedArgs)
+    {
+        CHECK_FAILURE(extendedArgs->put_ShouldPersist(FALSE));
+    }
+
+    // Do the rest asynchronously, to avoid calling dialog in an event handler.
+    m_appWindow->RunAsync([this, deferral, args]
+    {
+        COREWEBVIEW2_PERMISSION_KIND kind = COREWEBVIEW2_PERMISSION_KIND_UNKNOWN_PERMISSION;
+        BOOL userInitiated = FALSE;
+        wil::unique_cotaskmem_string uri;
+        CHECK_FAILURE(args->get_PermissionKind(&kind));
+        CHECK_FAILURE(args->get_IsUserInitiated(&userInitiated));
+        CHECK_FAILURE(args->get_Uri(&uri));
+
+        auto cachedKey = std::make_tuple(std::wstring(uri.get()), kind, userInitiated);
+        auto cachedPermission = m_cachedPermissions.find(cachedKey);
+        if (cachedPermission != m_cachedPermissions.end())
+        {
+            COREWEBVIEW2_PERMISSION_STATE state;
+            state =
+                (cachedPermission->second ? COREWEBVIEW2_PERMISSION_STATE_ALLOW
+                                           : COREWEBVIEW2_PERMISSION_STATE_DENY);
+            CHECK_FAILURE(args->put_State(state));
+        }
+        else
+        {
+            ShowPermissionDialog(uri, kind, userInitiated);
+        }
+        CHECK_FAILURE(deferral->Complete());
+    });
+    return S_OK;
+}
+```
+## SetPermission and GetNonDefaultPermissionCollection
+```c#
+List<CoreWebView2PermissionKind> _permissionKinds = new List<CoreWebView2PermissionKind>
+{
+  CoreWebView2PermissionKind.Microphone,
+  CoreWebView2PermissionKind.Camera,
+  CoreWebView2PermissionKind.Geolocation,
+  CoreWebView2PermissionKind.Notifications,
+  CoreWebView2PermissionKind.OtherSensors,
+  CoreWebView2PermissionKind.ClipboardRead,
+  CoreWebView2PermissionKind.AutomaticDownloads,
+  CoreWebView2PermissionKind.FileEditing,
+  CoreWebView2PermissionKind.Autoplay,
+  CoreWebView2PermissionKind.LocalFonts
+};
+
+List<CoreWebView2PermissionState> _permissionStates = new List<CoreWebView2PermissionState>
+{
+  CoreWebView2PermissionState.Allow,
+  CoreWebView2PermissionState.Deny,
+  CoreWebView2PermissionState.Default
+};
+
+// Gets the nondefault permission collection and updates a custom permission
+// management page.
+async void WebView_PermissionManager_DOMContentLoaded(object sender,
+    CoreWebView2DOMContentLoadedEventArgs arg)
+{
+    // Permission management APIs are only used on the app's permission
+    // management page.
+    if (webView.CoreWebView2.Source !=
+        "https://appassets.example/ScenarioPermissionManagement.html")
+    {
+        return;
+    }
+    // Gather all the nondefault permissions and post them to the app's
+    // permission management page. The permission management page can present
+    // a list of custom permissions set for this profile and let the end user
+    // modify them.
+    for (int i = 0; i < _permissionKinds.Count; i++)
```

Comment:
Its kind of clunky to force the end dev to make an array of all our enum values. Should we have a GetNonDefaultPermissionCollectionAsync that takes no parameters and gives you everything (no filtering by permission kind)?

Reply:
Or only the method that has no parameters and gives you everything? Your sample code doesn't want to apply filtering by permission kind. Do you think its still helpful to provide a version that does?

Reply:
I was thinking it could be helpful for building some sections of a permission management page. edge://settings/content has a section divided by permission kind, with subpages for each kind, like edge://settings/content/camera. Here's a developer ask: https://github.com/MicrosoftEdge/WebView2Feedback/issues/2089#issuecomment-1241893833

Reply:
I'll add a method without the filtering. Do you think I should keep this one too?

Reply:
Not sure. If you think it makes sense for your customers. Otherwise they should be able to do the filtering themselves and I would be fine removing the one with filtering